lib/src/main/kotlin/tree/AVLTree.kt

```diff
@@ -2,16 +2,142 @@ package tree
 
 import tree.node.AVLTreeNode
 
+
 class AVLTree<K : Comparable<K>, V> : SearchTree<K, V, AVLTreeNode<K, V>>() {
+
+    private fun height(node: AVLTreeNode<K, V>?): Int {
+        return node?.height ?: 0
+    }
+
+    private fun updateHeight(node: AVLTreeNode<K, V>?) {
+        if (node != null) {
+            node.height = 1 + maxOf(height(node.left), height(node.right))
+        }
+    }
+
+    private fun getBalance(node: AVLTreeNode<K, V>): Int {
+        return height(node.left) - height(node.right)
+    }
+
+    private fun rotateRight(node: AVLTreeNode<K, V>): AVLTreeNode<K, V>? {
+        val nodeLeft = node.left
+        val nodeLeftRight = nodeLeft?.right
+
+        nodeLeft?.right = node
+        node.left = nodeLeftRight
+
+        updateHeight(node)
+        updateHeight(nodeLeft)
+
+        return nodeLeft
+    }
+
+    private fun rotateLeft(node: AVLTreeNode<K, V>): AVLTreeNode<K, V>? {
+        val nodeRight = node.right
+        val nodeRightLeft = nodeRight?.left
+
+        nodeRight?.left = node
+        node.right = nodeRightLeft
+
+        updateHeight(node)
+        updateHeight(nodeRight)
+
+        return nodeRight
+    }
+
+    private fun rebalanced(root: AVLTreeNode<K, V>?, node: AVLTreeNode<K, V>?): AVLTreeNode<K, V>?{
+        updateHeight(root)
+        val balance = root?.let { getBalance(it) }
+        val rootLeft = root?.left
+        val rootRight = root?.right
+
+        if (node != null && balance != null){
+            if (balance > 1) {
+                if (rootLeft != null) {
+                    if (node.key < rootLeft.key) {
+                        return rotateRight(root)
+                    } else {
+                        root.left = root.left?.let { rotateLeft(it) }
+                        return rotateRight(root)
+                    }
+                }
+            }
+            if (balance < -1) {
+                if (rootRight != null) {
+                    if (node.key > rootRight.key) {
+                        return rotateLeft(root)
+                    } else {
+                        root.right = root.right?.let { rotateRight(it) }
+                        return rotateLeft(root)
+                    }
+                }
+            }
+        }
```

Comment:
Учитывая предыдущий комментарий код можно написать так:
Заметь, насколько мозгу проще это обрабатывать
И не забывай про новую строку между логическими блоками

```suggestion
        if (node != null && balance != null) {
            if (balance > 1 && rootLeft != null) {
                if (node.key < rootLeft.key) {
                    return rotateRight(root)
                }

                root.left = root.left?.let { rotateLeft(it) }
                return rotateRight(root)

            }
            
            if (balance < -1 && rootRight != null) {
                if (node.key > rootRight.key) {
                    return rotateLeft(root)
                }

                root.right = root.right?.let { rotateRight(it) }
                return rotateLeft(root)
            }
        }
```